# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [1]:
print('Escola de Dados da Alura!')

Escola de Dados da Alura!


### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [2]:
print('Nome: Jefferson \nSobrenome: Campos')

Nome: Jefferson 
Sobrenome: Campos


### Questão 3
Imprima o seu primeiro nome letra a letra. Por exemplo, nome é Mirla, então eu obtenho a seguinte saída:
```
M
I
R
L
A
```


In [3]:
print('J')

J


In [4]:
print('e')

e


In [5]:
print('f')

f


In [6]:
print('f')

f


In [7]:
print('e')

e


In [8]:
print('r')

r


In [9]:
print('s')

s


In [10]:
print('o')

o


In [11]:
print('n')

n


In [12]:
print('J\ne\nf\nf\ne\nr\ns\no\nn')

J
e
f
f
e
r
s
o
n


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [14]:
print(5, 'Agosto', 1986)

5 Agosto 1986


### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [15]:
print(2026)

2026


## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [1]:
name: str = input('Inform your name: ')
print(f'Hello, {name}')

Inform your name:  Jeff


Hello, Jeff


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [3]:
name: str = input('Inform your name: ')
age: int = int(input('Inform your age: '))
print(f'Hello, {name}, you are {age} year old.')

Inform your name:  Jeff
Inform your age:  39


Hello, Jeff, you are 39 year old.


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [5]:
name: str = input('Inform your name: ')
age: int = int(input('Inform your age: '))
height: float = float(input('Inform your height: '))
print(f'Hello, {name}, you are {age} year old and has {height} of height.')

Inform your name:  Jeff
Inform your age:  39
Inform your height:  1.70


Hello, Jeff, you are 39 year old and has 1.7 of height.


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [6]:
number_01: int = int(input('Inform a number 01: '))
number_02: int = int(input('Inform a number 02: '))
total: int = number_01 + number_02
print(f'Total for {number_01} + {number_02} = {total}')

Inform a number 01:  5
Inform a number 02:  8


Total for 5 + 8 = 13


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [7]:
number_01: int = int(input('Inform a number 01: '))
number_02: int = int(input('Inform a number 02: '))
number_03: int = int(input('Inform a number 03: '))
total: int = number_01 + number_02 + number_03
print(f'Total for {number_01} + {number_02} + {number_03} = {total}')

Inform a number 01:  1
Inform a number 02:  6
Inform a number 03:  2


Total for 1 + 6 + 2 = 9


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [8]:
number_01: int = int(input('Inform a number 01: '))
number_02: int = int(input('Inform a number 02: '))
total: int = number_01 - number_02
print(f'Total for {number_01} - {number_02} = {total}')

Inform a number 01:  5
Inform a number 02:  2


Total for 5 - 2 = 3


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [9]:
number_01: int = int(input('Inform a number 01: '))
number_02: int = int(input('Inform a number 02: '))
total: int = number_01 * number_02
print(f'Total for {number_01} * {number_02} = {total}')

Inform a number 01:  5
Inform a number 02:  9


Total for 5 * 9 = 45


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [14]:
number_01: int = int(input('Inform a number 01: '))
number_02: int = int(input("Inform a number 02 (can't be 0 - zero): "))
if number_02 == 0:
    raise Exception("Number 02 can't be zero!")
total: int = number_01 / number_02
print(f'Total for {number_01} / {number_02} = {total}')

Inform a number 01:  9
Inform a number 02:  3


Total for 9 / 3 = 3.0


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [15]:
number_01: int = int(input('Inform a number 01: '))
number_02: int = int(input('Inform a number 02: '))
if number_02 == 0:
    raise Exception("Number 02 can't be zero!")
total: int = number_01 ** number_02
print(f'Total for {number_01} ** {number_02} = {total}')

Inform a number 01:  2
Inform a number 02:  3


Total for 2 ** 3 = 8


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [18]:
number_01: int = int(input('Inform a number 01: '))
number_02: int = int(input("Inform a number 02 (can't be 0 - zero): "))
if number_02 == 0:
    raise Exception("Number 02 can't be zero!")
total: int = number_01 // number_02
print(f'Total for {number_01} // {number_02} = {total}')

Inform a number 01:  15
Inform a number 02 (can't be 0 - zero):  8


Total for 15 // 8 = 1


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [21]:
number_01: int = int(input('Inform a number 01: '))
number_02: int = int(input("Inform a number 02 (can't be 0 - zero): "))
if number_02 == 0:
    raise Exception("Number 02 can't be zero!")
total: int = number_01 % number_02
print(f'Total for {number_01} % {number_02} = {total}')

Inform a number 01:  9
Inform a number 02 (can't be 0 - zero):  5


Total for 9 % 5 = 4


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [23]:
grade_01: int = int(input('Inform a grade 01: '))
grade_02: int = int(input('Inform a grade 02: '))
grade_03: int = int(input('Inform a grade 03: '))
avg: int = (grade_01 + grade_02 + grade_03) / 3
print(f'Total for avarage = {avg}')

Inform a grade 01:  10
Inform a grade 02:  8
Inform a grade 03:  9


Total for avarage = 9.0


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [29]:
grade_01: int = int(input('Inform a grade 01: '))
grade_02: int = int(input('Inform a grade 02: '))
grade_03: int = int(input('Inform a grade 03: '))
grade_04: int = int(input('Inform a grade 04: '))
avg: int = ((grade_01 * 1) + (grade_02 * 2) + (grade_03 * 3) + (grade_04 * 4)) / 10
print(f'Total for avarage = {avg}')
avg_pond: int = (5 * 1 + 12 * 2 + 20 * 3 + 15 * 4) / (1 + 2 + 3 + 4)
print(f'{avg_pond=}')

Inform a grade 01:  5
Inform a grade 02:  12
Inform a grade 03:  20
Inform a grade 04:  15


Total for avarage = 14.9
avg_pond=14.9


### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [33]:
frase: str = 'Hello, world!'
print(frase)

Hello, world!


#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [34]:
frase: str = input('Inform your phrase: ')
print(frase)

Inform your phrase:  Hello, world!!


Hello, world!!


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [35]:
phrase: str = input('Inform your phrase: ')
print(f'original: {phrase} ::: upper: {phrase.upper()}')

Inform your phrase:  Hello, world!!!


original: Hello, world!!! ::: upper: HELLO, WORLD!!!


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [37]:
phrase: str = input('Inform your phrase: ')
print(f'original: {phrase} ::: lower: {phrase.lower()}')

Inform your phrase:  Hello, World!!!


original: Hello, World!!! ::: lower: hello, world!!!


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [39]:
phrase: str = '      Hello, World!               '
print(f'original: {phrase} ::: strip: {phrase.strip()}')

original:       Hello, World!                ::: strip: Hello, World!


#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [40]:
phrase: str = input('Inform your phrase: ')
print(f'original: {phrase} ::: strip: {phrase.strip()}')

Inform your phrase:     Hello, world!!     


original:    Hello, world!!      ::: strip: Hello, world!!


#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [41]:
phrase: str = input('Inform your phrase: ')
print(f'original: {phrase} ::: strip: {phrase.strip().lower()}')

Inform your phrase:     Hello, World!!         


original:    Hello, World!!          ::: strip: hello, world!!


#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [42]:
phrase: str = input('Inform your phrase: ')
print(f'original: {phrase} ::: strip: {phrase.lower().replace("e", "f")}')

Inform your phrase:  Hello, World!!


original: Hello, World!! ::: strip: hfllo, world!!


#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [46]:
phrase: str = input('Inform your phrase: ')
print(f'original: {phrase} ::: strip: {phrase.lower().replace("a", chr(64))}')

Inform your phrase:  Hello, World!! Amiga PC!


original: Hello, World!! Amiga PC! ::: strip: hello, world!! @mig@ pc!


#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [47]:
phrase: str = input('Inform your phrase: ')
print(f'original: {phrase} ::: strip: {phrase.lower().replace("s", chr(36))}')

Inform your phrase:  Hello, World!! Amiga PC! Sum it up untl the end as you saw it!


original: Hello, World!! Amiga PC! Sum it up untl the end as you saw it! ::: strip: hello, world!! amiga pc! $um it up untl the end a$ you $aw it!


## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [1]:
n1: int = int(input('Inform number 01: '))
n2: int = int(input('Inform number 02: '))
bigger: int
if n1 > n2:
    bigger = n1
else:
    bigger = n2

print(f'Bigger number ({n1}, {n2}) is: {bigger} (max({n1}, {n2}) is {max(n1, n2)})')

Inform number 01:  10
Inform number 02:  10


Bigger number (10, 10) is: 10 (max(10, 10) is 10)


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [7]:
result: float = float(input('Inform results for this quarter: '))
status: str = ''
if result > 0:
    status = 'POSITIVE'
elif result == 0:
    status = 'NEUTRAL'
else:
    status = 'NEGATIVE'

print(f'Result {result} for this quarter is: {status}')

Inform results for this quarter:  0


Result 0.0 for this quarter is: NEUTRAL


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [9]:
vowels: List[str] = ['a', 'e', 'i', 'o', 'u']
letter: str = input('Inform a letter: ')
status: bool = False
if letter in vowels:
    status = True

print(f'Letter {letter} is vowel: {status}')

Inform a letter:  e


Letter e is vowel: True


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [21]:
p1: float = float(input('Inform price for year 01: '))
p2: float = float(input('Inform price for year 02: '))
p3: float = float(input('Inform price for year 03: '))
p_max: float
p_min: float
# chained
if p1 > p2:
    if p1 > p3:
        p_max = p1
    else:
        p_max = p3
else:
    if p2 > p3:
        p_max = p2
    else:
        p_max = p3

if p1 < p2:
    if p1 < p3:
        p_min = p1
    else:
        p_min = p3
else:
    if p2 < p3:
        p_min = p2
    else:
        p_min = p3

greater: float = p1
if p2 > greater:
    greater = p2
if p3 > greater:
    greater = p3

lower: float = p1
if p2 < lower:
    lower = p2
if p3 < lower:
    lower = p3

print(f'{greater=} {lower=} medium={[p for p in [p1, p2, p3] if p not in [min(p1, p2, p3), max(p1, p2, p3)]]}')
print(f'Result for prices: price 01 (${p1}); price 02 (${p2}); price 03 (${p3}); price minimum: {p_min} price maximum: ${p_max} (min(p1, p2, p3) is {min(p1, p2, p3)}; max(p1, p2, p3) is {max(p1, p2, p3)})')

Inform price for year 01:  100
Inform price for year 02:  50
Inform price for year 03:  300


greater=300.0 lower=50.0 medium=[100.0]
Result for prices: price 01 ($100.0); price 02 ($50.0); price 03 ($300.0); price minimum: 50.0 price maximum: $300.0 (min(p1, p2, p3) is 50.0; max(p1, p2, p3) is 300.0)


#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [20]:
p1: float = float(input('Inform price for product 01: '))
p2: float = float(input('Inform price for product 02: '))
p3: float = float(input('Inform price for product 03: '))

p_min: float
if p1 < p2:
    if p1 < p3:
        p_min = p1
    else:
        p_min = p3
else:
    if p2 < p3:
        p_min = p2
    else:
        p_min = p3

print(f'Result for prices: price 01 (${p1}); price 02 (${p2}); price 03 (${p3}); price minimum: {p_min} (min(p1, p2, p3) is {min(p1, p2, p3)})')

Inform price for product 01:  100
Inform price for product 02:  200
Inform price for product 03:  300


Result for prices: price 01 ($100.0); price 02 ($200.0); price 03 ($300.0); price minimum: 100.0 (min(p1, p2, p3) is 100.0)


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [23]:
p1: float = float(input('Inform number 01: '))
p2: float = float(input('Inform number 02: '))
p3: float = float(input('Inform number 03: '))

lowerest: float
medium: float
bigger: float

if p1 < p2 and p1 < p3:
    lowerest = p1
elif p2 < p1 and p2 < p3:
    lowerest = p2
elif p3 < p1 and p3 < p2:
    lowrest = p3

if p1 > p2 and p1 > p3:
    bigger = p1
elif p2 > p1 and p2 > p3:
    bigger = p2
elif p3 > p2 and p3 > p2:
    bigger = p3

if p1 != lowerest and p1 != bigger:
    medium = p1
elif p2 != lowerest and p2 != bigger:
    medium = p2
elif p3 != lowerest and p3 != bigger:
    medium = p3

print(f'Result for prices: price 01 (${p1}); price 02 (${p2}); price 03 (${p3});\n lowerest: {lowerest}\n medium: {medium}\n bigger: {bigger}\n (min(p1, p2, p3) is {min(p1, p2, p3)}\n; max(p1, p2, p3) {max(p1, p2, p3)}\n; medium -> p1 {p1} in [min(p1, p2, p3) {min(p1, p2, p3)}, max(p1, p2, p3) {max(p1, p2, p3)}]: {p1 in [min(p1, p2, p3), max(p1, p2, p3)]} :: p2 {p2} in [min(p1, p2, p3) {min(p1, p2, p3)}, max(p1, p2, p3) {max(p1, p2, p3)}]: {p2 in [min(p1, p2, p3), max(p1, p2, p3)]} :: p3 {p3} in [min(p1, p2, p3) {min(p1, p2, p3)}, max(p1, p2, p3) {max(p1, p2, p3)}]: {p3 in [min(p1, p2, p3), max(p1, p2, p3)]})')

Inform number 01:  100
Inform number 02:  200
Inform number 03:  300


Result for prices: price 01 ($100.0); price 02 ($200.0); price 03 ($300.0);
 lowerest: 100.0
 medium: 200.0
 bigger: 300.0
 (min(p1, p2, p3) is 100.0
; max(p1, p2, p3) 300.0
; medium -> p1 100.0 in [min(p1, p2, p3) 100.0, max(p1, p2, p3) 300.0]: True :: p2 200.0 in [min(p1, p2, p3) 100.0, max(p1, p2, p3) 300.0]: False :: p3 300.0 in [min(p1, p2, p3) 100.0, max(p1, p2, p3) 300.0]: True)


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [36]:
shift: str = input(f'Inform your shift: [Morning | Afternoon | Nightly]: ')
msg: str = ''

if shift.lower() == 'morning':
    msg = 'GOOD MORNING!!'
elif shift.lower() == 'afternoon':
    msg = 'GOOD AFTERNOON!!'
elif shift.lower() == 'nightly':
    msg = 'GOOD NIGHT!!'
else:
    msg = 'Invalid Value!!'

print(msg)

Inform your shift: [Morning | Afternoon | Nightly]:  madrugada


Invalid Value!!


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [38]:
n1: float = float(input('Inform a number: '))
if (n1 % 2) == 0:
    print(f'Number {n1} is EVEN')
else:
    print(f'Number {n1} is ODD')

Inform a number:  11


Number 11.0 is ODD


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [27]:
n1: float = float(str(input('Inform a number: ')))

# alternative
if (n1 - int(n1)) == 0:
    print(f'Number {n1} is INT --> {n1 - int(n1)}')
else:
    print(f'Number {n1} is FLOAT --> {n1 - int(n1)}')

if (n1 % 1) == 0:
    print(f'Number {n1} is INT --> {n1 % 1}')
else:
    print(f'Number {n1} is FLOAT --> {n1 % 1}')

Inform a number:  3.3


Number 3.3 is FLOAT --> 0.2999999999999998
Number 3.3 is FLOAT --> 0.2999999999999998


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [30]:
n1: float = float(input('Inform number 01: '))
n2: float = float(input('Inform number 02: '))
action: str = input('Inform action [+, -, *, **, /, //, %]: ')
total: float
if action == '+':
    total = n1 + n2
elif action == '-':
    total = n1 - n2
elif action == '*':
    total = n1 * n2
elif action == '**':
    total = n1 ** n2
elif action == '/':
    total = n1 / n2
elif action == '//':
    total = n1 // n2
elif action == '%':
    total = n1 % n2
else:
    raise Exception(f'Invlaid action: {action}')

if (total % 2) == 0:
    print(f'total: {total} is EVEN')
else:
    print(f'total: {total} is ODD')

if total >= 0:
    print(f'total: {total} is POSITIVE (zero will be considered positive too!!)')
else:
    print(f'total: {total} is NEGATIVE')

if (total - int(total)) == 0:
    print(f'Number {total} is INT --> left over {total - int(total)}')
else:
    print(f'Number {total} is FLOAT --> left over {total - int(total)}')

if (total % 1) == 0:
    print(f'Number {total} is INT --> left over {total % 1}')
else:
    print(f'Number {total} is FLOAT --> left over {total % 1}')
    

Inform number 01:  10
Inform number 02:  3
Inform action [+, -, *, **, /, //, %]:  /


total: 3.3333333333333335 is ODD
total: 3.3333333333333335 is POSITIVE (zero will be considered positive too!!)
Number 3.3333333333333335 is FLOAT --> left over 0.3333333333333335
Number 3.3333333333333335 is FLOAT --> left over 0.3333333333333335


#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dois lados iguais;
- Triângulo Escaleno: três lados diferentes;


In [35]:
l1: float = float(input('Inform side 01: '))
l2: float = float(input('Inform side 02: '))
l3: float = float(input('Inform side 03: '))
valid: bool = False
triangle_type: str = ''
if (l1 + l2) > l3 and (l1 + l3) > l2 and (l2 + l3) > l1:
    valid = True

if valid:
    if l1 == l2 == l3:
        triangle_type = 'equilateral'
    elif l1 == l2 or l1 == l3 or l2 == l3:
        triangle_type = 'isosceles'
    else:
        triangle_type = 'scalene'

print(f'l1: {l1} l2: {l2} l3: {l3} -> valid: {valid} :: type {triangle_type}')

Inform side 01:  4
Inform side 02:  7
Inform side 03:  10


l1: 4.0 l2: 7.0 l3: 10.0 -> valid: True :: type scalene


#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [16]:
total: float = float(input('Inform the total to fill: '))
gas: str = input('Inform the gas type: E - Etanol | D - Diesel')

price: float = 0.0
if gas == 'E':
    price = 1.7
else:
    price = 2.0

discount_rate: float = 0.0
if gas == 'E':
    if total <= 15:
        discount_rate = 0.02
    else:
        discount_rate = 0.03
else:
    if total <= 15:
        discount_rate = 0.04
    else:
        discount_rate = 0.05

full_price: float = (price * total)
discount: float = full_price * (discount_rate)
final_price: float = full_price - discount

print(f'{gas=} {total=} {price=} {discount_rate=} {full_price=} {discount=} {final_price=}')

Inform the total to fill:  20
Inform the gas type: E - Etanol | D - Diesel E


gas='E' total=20.0 price=1.7 discount_rate=0.03 full_price=34.0 discount=1.02 final_price=32.98


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


In [28]:
y1: float = float(input('Inform the total for year 01: '))
y2: float = float(input('Inform the total for year 02: '))
result: float = y2 - y1
fluctuation: float = result / y1

recomendation: str = 'DO NOTHING'
if fluctuation >= 0.2:
    recomendation = 'FULL BONUS SALES TEAM'
elif fluctuation >= 0.02:
    recomendation = 'SMALL BONUS SALES TEAM'
elif fluctuation >= -0.1:
    recomendation = 'PLANNING NEW SALES INCENTIVE POLICES'
else:
    recomendation = 'COST CUTTING'

print(f'{y1=} {y2=} {result=} {fluctuation=} {recomendation=}')

Inform the total for year 01:  100
Inform the total for year 02:  95


y1=100.0 y2=95.0 result=-5.0 fluctuation=-0.05 recomendation='PLANNING NEW SALES INCENTIVE POLICES'


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

In [4]:
n1: int = int(input('Inform your the start number >: '))
n2: int = int(input('Inform your the final number >: '))
for i in range(n1, n2 + 1, 1):
    print(i)

Inform your the start number >:  1
Inform your the final number >:  5


1
2
3
4
5


#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

In [10]:
a: int = 4
b: int = 10
day: int = 0
while a < b:
    a += a * 1.03
    b += b * 1.015
    day += 1
    print(f'day {day} a: {a:.2f} b: {b:.2f} % {(a / b) * 100}')

print(f'Took {day} day(s) to a ({a:.2f} overrun b {b:.2f}!!')

day 1 a: 8.12 b: 20.15 % 40.29776674937966
day 2 a: 16.48 b: 40.60 % 40.59775012468522
day 3 a: 33.46 b: 81.81 % 40.89996662685409
day 4 a: 67.93 b: 164.85 % 41.20443287965946
day 5 a: 137.89 b: 332.18 % 41.51116563062467
day 6 a: 279.92 b: 669.35 % 41.82018175194446
day 7 a: 568.24 b: 1348.73 % 42.13149824141303
day 8 a: 1153.53 b: 2717.69 % 42.44513222335904
day 9 a: 2341.66 b: 5476.15 % 42.76110094958752
day 10 a: 4753.58 b: 11034.45 % 43.07942180032887
day 11 a: 9649.76 b: 22234.41 % 43.40011228519485
day 12 a: 19589.01 b: 44802.33 % 43.72319004414171
day 13 a: 39765.69 b: 90276.70 % 44.04867284844053
day 14 a: 80724.34 b: 181907.54 % 44.37657860165473
day 15 a: 163870.42 b: 366543.70 % 44.70692534062486
day 16 a: 332656.95 b: 738585.55 % 45.03973123646078
day 17 a: 675293.60 b: 1488249.89 % 45.37501459554114
day 18 a: 1370846.01 b: 2998823.52 % 45.71279386052035
day 19 a: 2782817.41 b: 6042629.40 % 46.05308761134309
day 20 a: 5649119.34 b: 12175898.23 % 46.395914566266235
day 21 a

#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

In [13]:
for i in range(1, 16):
    grade: int = -1
    while grade < 0 or grade > 5:
        grade = int(input(f'Inform your grade #{i}: '))
    print(f'Your grade #{i} {grade} is correct!!')

Inform your grade #1:  1


Your grade #1 1 is correct!!


Inform your grade #2:  5


Your grade #2 5 is correct!!


Inform your grade #3:  3


Your grade #3 3 is correct!!


Inform your grade #4:  9
Inform your grade #4:  9
Inform your grade #4:  7
Inform your grade #4:  5


Your grade #4 5 is correct!!


Inform your grade #5:  2


Your grade #5 2 is correct!!


Inform your grade #6:  -2
Inform your grade #6:  -9
Inform your grade #6:  0


Your grade #6 0 is correct!!


Inform your grade #7:  1


Your grade #7 1 is correct!!


Inform your grade #8:  5


Your grade #8 5 is correct!!


Inform your grade #9:  6
Inform your grade #9:  5


Your grade #9 5 is correct!!


Inform your grade #10:  1


Your grade #10 1 is correct!!


Inform your grade #11:  1


Your grade #11 1 is correct!!


Inform your grade #12:  2


Your grade #12 2 is correct!!


Inform your grade #13:  3


Your grade #13 3 is correct!!


Inform your grade #14:  4


Your grade #14 4 is correct!!


Inform your grade #15:  5


Your grade #15 5 is correct!!


#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

In [20]:
temperatures: List[float] = []
while True:
    t: float = float(input('Inform your temperature: '))
    if t == -273:
        break
    temperatures.append(t)
    
print(f'Avarage of temperatues ({len(temperatures)} total): {sum(temperatures) / len(temperatures)} ºC--> {"; ".join([f"{te}ºC" for te in temperatures])}')

Inform your temperature:  10
Inform your temperature:  20
Inform your temperature:  -273


Avarage of temperatues (2 total): 15.0 ºC--> 10.0ºC; 20.0ºC


#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

In [39]:
n1: int = int(input('Inform your number >: '))
i = n1
total: int = 1
while i > 0:
    total *= i
    i -= 1

print(f'Final factorial: {total}')

Inform your number >:  7


Final factorial: 5040


### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

In [42]:
n: int = int(input('Inform the number for multiplation table>: '))
for i in range(1, 11):
    print(f'{n} X {i} = {n * i}')

Inform the number for multiplation table>:  5


5 X 1 = 5
5 X 2 = 10
5 X 3 = 15
5 X 4 = 20
5 X 5 = 25
5 X 6 = 30
5 X 7 = 35
5 X 8 = 40
5 X 9 = 45
5 X 10 = 50


#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

In [3]:
n: int = int(input('Inform the number for multiplation table>: '))
is_prime: bool = True
first_divisor: int = 0
for i in range(2, n):
    if n % i == 0:
        first_divisor = i
        is_prime = False
        break

print(f'Number {n} is prime: {is_prime} (first divisor found (if have): {first_divisor})')

Inform the number for multiplation table>:  107


Number 107 is prime: True (first divisor found (if have): 0)


#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

In [64]:
ages: List[int] = []
# while True:
#     age: int = int(input('Inform the age>: '))
#     if age < 0:
#         break
#     ages.append(age)

ages = [
    99, 63, 47, 22,
    85, 56, 33, 17,
    76, 70, 44, 15
]

ages_000_025: List[int] = [a for a in ages if 0 <= a <= 25]
ages_026_050: List[int] = [a for a in ages if 26 <= a <= 50]
ages_051_075: List[int] = [a for a in ages if 51 <= a <= 75]
ages_076_100: List[int] = [a for a in ages if 76 <= a <= 100]
print(f'ages_000_025: {len(ages_000_025)} ({round((len(ages_000_025) / len(ages)) * 100, 2)}%: {ages_000_025}')
print(f'ages_026_050: {len(ages_026_050)} ({round((len(ages_026_050) / len(ages)) * 100, 2)}%: {ages_026_050}')
print(f'ages_051_075: {len(ages_051_075)} ({round((len(ages_051_075) / len(ages)) * 100, 2)}%: {ages_051_075}')
print(f'ages_076_100: {len(ages_076_100)} ({round((len(ages_076_100) / len(ages)) * 100, 2)}%: {ages_076_100}')

ages_000_025: 3 (25.0%: [22, 17, 15]
ages_026_050: 3 (25.0%: [47, 33, 44]
ages_051_075: 3 (25.0%: [63, 56, 70]
ages_076_100: 3 (25.0%: [99, 85, 76]


#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

In [72]:
pool: List[int] = [
    1, 2, 3, 4, 5,
    1, 2, 2, 5, 6,
    1, 2, 3, 6, 6,
    1, 2, 3, 2, 6
]
for i in range(1, 7):
    votes: List[int] = [v for v in pool if v == i]
    print(f'for candidate #{i}: {len(votes)} vote(s) ({round((len(votes) / len(pool)) * 100, 2):05.2f})%{" (null vote)" if i == 5 else " (white)" if i == 6 else ""}')

for candidate #1: 4 vote(s) (20.00)%
for candidate #2: 6 vote(s) (30.00)%
for candidate #3: 3 vote(s) (15.00)%
for candidate #4: 1 vote(s) (05.00)%
for candidate #5: 2 vote(s) (10.00)% (null vote)
for candidate #6: 4 vote(s) (20.00)% (white)


## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

In [9]:
expenses: List = [2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]
avg: float = sum(expenses) / len(expenses)
print(f'{avg}')
print(f'Total expenses: ${round(avg, 2)}')

3144.222
Total expenses: $3144.22


#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

In [15]:
expenses: List = [2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]
threshold: List[float] = [expense for expense in expenses if expense > 3000.00]
print(f'Total expense units  > $3000.00: {len(threshold)} of {len(expenses)} ({(len(threshold) / len(expenses)) * 100}%)')
print(f'Total expense values > $3000.00: ${sum(threshold):.2f} of ${sum(expenses):.2f} ({((sum(threshold) / sum(expenses)) * 100):.2f}%)')

Total expense units  > $3000.00: 7 of 10 (70.0%)
Total expense values > $3000.00: $24111.22 of $31442.22 (76.68%)


#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

In [16]:
values: List[int] = []
for i in range(1, 6):
    values.append(int(input('Inform an integer number>: ')))

print(values)

Inform an integer number>:  1
Inform an integer number>:  5
Inform an integer number>:  99
Inform an integer number>:  3573
Inform an integer number>:  12


[1, 5, 99, 3573, 12]


#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

In [19]:
values: List[int] = []
for i in range(1, 6):
    values.append(int(input('Inform an integer number>: ')))

values.reverse()
print(values)

Inform an integer number>:  1
Inform an integer number>:  5
Inform an integer number>:  9
Inform an integer number>:  11
Inform an integer number>:  35


[35, 11, 9, 5, 1]


#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

In [27]:
num: int = int(input('Inform a number >: '))
primes: List[int] = []
for i in range(2,  num + 1):
    is_prime: bool = True
    for j in range(2, i):
        if i % j == 0:
            first_divisor = j
            is_prime = False
            break
    if is_prime:
        primes.append(i)

print(primes)

Inform a number >:  33


[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]


#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

In [31]:
invalid = False
my_date = {}
my_date['day'] = int(input('Inform your day >: '))
my_date['month'] = int(input('Inform your month >: '))
my_date['year'] = int(input('Inform your year >: '))

if 1 < my_date['day'] > 30:
    invalid = True
    print('invalid day')

if 1 < my_date['month'] > 12:
    invalid = True
    print('invalid month')

if 1 < my_date['year'] > 3000:
    invalid = True
    print('invalid year')

if not invalid:
    print(f'{my_date}')


Inform your day >:  10
Inform your month >:  13
Inform your year >:  2025


invalid month


### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
``` 
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.
